### PR TITLE
Add a cheatsheet

### PR DESCRIPTION
**Pull request by DjebbZ**
First draft of a cheatsheet. Far from finished. I put all of Specter API from the Wiki pages (List of navigators and macros).

I used a bit more categories and subcategories than we talked in https://github.com/nathanmarz/specter/issues/56. But the categorization is a draft. And there is a `Don't know yet` category for things I didn't know where to put.

Looking forward to your review!

### CODE REVIEW

File: Cheat-Sheet.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+`select-first`, `select-one`, `select-one!`
+
+###### Query test
+
+`selected-any?`
+
+###### Transducer-related
+
+`traverse`, `traverse-all`
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: `traverse` is not transducer-related

File: Cheat-Sheet.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+## Operations
+
+There are 2 types of operations: queries and transforms.
+
+#### Query
+
+###### Multiple elements (return a vector of values found)
+
+`select`, `select-any`
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: `select-any` returns a single value

File: Cheat-Sheet.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+`INDEXED-VALS`, `before-index`, `index-nav`, `nthpath`, `srange`, `srange-dynamic`
+
+###### Others
+
+`continuous-subseqs`
+
+#### Sets
+
+`NONE-ELEM`, `set-elem`
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: missing `subset`

File: Cheat-Sheet.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+#### Keywords/Symbols
+
+`NAME`, `NAMESPACE`
+
+#### Atoms
+
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: missing `ATOM`

File: Cheat-Sheet.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+`regex-nav`, `srange`
+
+#### Metadata
+
+`ALL-WITH-META`, `META`
+
+#### Views
+
+`NIL->LIST`, `NIL->SET`, `NIL->VECTOR`, `filterer`, `nil->val`, `subselect`, `transform`, `traversed`, `view`
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: `filterer` should go under Sequence

Comment (nathanmarz) [reply Cheat-Sheet.md]: `transformed`, not `transform`

File: Cheat-Sheet.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+#### Value collection
+
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: missing `DISPENSE`

File: Cheat-Sheet.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+`VAL`, `collect`, `collect-one`, `collected?`, `putval`, `with-fresh-collected`
+
+#### Don't know yet
+
+`NONE`, `codewalker`, `each-nav`, `multi-path`, `must`, `parser`, `walker`
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: `NONE` is not a navigator and should be removed, `eachnav` belongs under "Custom Navigators", `must` belongs under "Maps", `parser` fits best under "Views", `multi-path` belongs under "Control", and `walker`/`codewalker` should get their own section "Walking"

File: Cheat-Sheet.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+#### Multi-transform
+
+`terminal`, `terminal-val`
+
+## Extending Specter
+
+#### Custom paths
+
+`declarepath`, `defprotocolpath`, `extend-protocolpath`, `path`, `providepath`, `recursive-path`
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: missing `local-declarepath`

**Comment by nathanmarz** [comment]
Made inline comments. More generally, I think there's over-categorization here. A cheatsheet should just be a quick reference based on broad criteria. I think this should put all the query operations in one category, all the sequence operations into one category, all the custom navigator operations into one category, etc.

Also, this needs every API element to link to either the API docs or the wiki pages to be useful.

**Comment by DjebbZ** [comment]
Fixed almsot all categorization issues. Remains `traverse`, still not sure where it should beloing. In `Sequences` ?

Regarding links, sure it was part of the plan, I just wanted to get categorization right first, because links pollute the reading of the PR. Will do it after your next review.

**Comment by DjebbZ** [comment]
Also, I plan to align all function names/links horizontally not vertically, so it looks more like a usual cheatsheet (which is often 2-dimensional).

**Comment by DjebbZ** [comment]
Normally it's all good. I put `traverse` in `Query`, added links everywhere and manually verified all of them. I also fixed several links/anchors in the current wiki pages in the process.

Nothing beats a review though.

Quite happy with the current shape of it, even though I just realised there's no mention of the `compiled-*` version of the API in the cheat sheet. I think there should be.

**Comment by DjebbZ** [comment]
Now that I added the compiled versions I think we're _really_ all good.

File: Cheat-Sheet.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+#### Control
+
+[`STAY`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#stay), [`STOP`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#stop), [`comp-paths`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#comp-paths), [`cond-path`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#cond-path), [`continue-then-stay`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#continue-then-stay), [`if-path`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#if-path), [`multi-path`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#multi-path), [`stay-then-continue`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#stay-then-continue)
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: `comp-paths` is not a navigator and should be removed.

File: Cheat-Sheet.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+#### Walking
+
+[`codewalker`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#codewalker), [`walker`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#walker)
+
+#### Multi-transform
+
+[`terminal`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal), [`terminal-val`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal-val)
+
+## Extending Specter
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: These three categories should be merged and named "Custom navigators"

File: Cheat-Sheet.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+#### Metadata
+
+[`ALL-WITH-META`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#all-with-meta), [`META`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#meta) 
+
+#### Views
+
+[`NIL->LIST`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-list), [`NIL->SET`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-set), [`NIL->VECTOR`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-vector), [`nil->val`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-val), [`parser`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#parser), [`subselect`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#subselect), [`transformed`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#transformed), [`traversed`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#traversed), [`view`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#view)
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: `subselect` doesn't belong here. Probably the best place to put it is "Control".

File: Cheat-Sheet.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+There are 2 types of operations: queries and transforms. Most of them have a compiled version that uses precompiled paths.
+
+#### Query
+
+[`select`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#select), [`select-any`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#select-any), [`select-first`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#select-first), [`select-one`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#select-one), [`select-one!`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#select-one-1), [`selected-any?`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#selected-any), [`traverse`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#traverse), [`traverse-all`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#traverse-all)
+
+#### Transform
+
+[`transform`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#transform), [`multi-transform`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#multi-transform), [`replace-in`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#replace-in), [`setval`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#setval)
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: Should add `vtransform`

File: Cheat-Sheet.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+[`pred`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#pred), [`pred=`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#pred-1), [`pred<`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#pred-2), [`pred>`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#pred-3), [`pred<=`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#pred-4), [`pred>=`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#pred-5), [`not-selected?`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#not-selected),  [`selected?`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#selected)
+
+#### Walking
+
+[`codewalker`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#codewalker), [`walker`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#walker)
+
+#### Multi-transform
+
+[`terminal`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal), [`terminal-val`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal-val)
```
Comment (nathanmarz) [inline Cheat-Sheet.md]: Should add `vterminal`

**Comment by nathanmarz** [comment]
Looks really good overall. Made more inline comments.

Agree that changing the formatting to be more like a usual cheatsheet would be a big improvement.

I don't think the compiled-* operations belong in the cheatsheet as they are pretty much never needed.

One last navigator missing from the list is `compact`. Probably fits best being listed under "Maps", "Sequences", and "Sets".

File: Cheat-Sheet.md (hunk 9 above)
Comment (DjebbZ) [reply Cheat-Sheet.md]: Where should it go then ? Under the new  `Compiled` section in `Operations` ?

File: Cheat-Sheet.md (hunk 11 above)
Comment (DjebbZ) [reply Cheat-Sheet.md]: Seems strange to me since the wiki says it's a view, but, ok...

File: Cheat-Sheet.md (hunk 12 above)
Comment (DjebbZ) [reply Cheat-Sheet.md]: Missed that since it's not in the Wiki.

File: Cheat-Sheet.md (hunk 13 above)
Comment (DjebbZ) [reply Cheat-Sheet.md]: Missed that since it's not in the Wiki.

File: Cheat-Sheet.md (hunk 10 above)
Comment (DjebbZ) [reply Cheat-Sheet.md]: Ok

File: Cheat-Sheet.md (hunk 9 above)
Comment (DjebbZ) [reply Cheat-Sheet.md]: Since you want to remove the `compiled-*` version this one should be removed too, got it.

**Comment by DjebbZ** [comment]
Fixed all the points you mentionned, apart from 

> Agree that changing the formatting to be more like a usual cheatsheet would be a big improvement.

Do you mean that the formatting currently in the PR is the improvement or that it needs to be improved ? I basically put categories one under the other (normal document flow) and put each API item in the same line (horizontally). This is what I meant by "improvement". Are you saying you'd like something else ?

**Comment by DjebbZ** [comment]
Realized I forgot to add `local-declarepath`.

**Comment by DjebbZ** [comment]
For this PR to be really complete, you @nathanmarz need to add documentation for the API that isn't in the wiki and link to it from the Cheat Sheet. I believe you're the only one able to do it. I looked at the changelog, there are hints as to how to document them but I'm not sure copy-pasting the changelog would be enough.

**Comment by nathanmarz** [comment]
This is great, thanks. I also linked to it from the README. I made a note at https://github.com/nathanmarz/specter-wiki/issues/17 to document those last few things.

**Comment by DjebbZ** [comment]
You're welcome. By the way I'm curious about how you merged it in the real wiki, since it's a different repository. Did you just set `specter.wiki` as a different remote in git ?

**Comment by nathanmarz** [comment]
Yes, that's exactly what I have set up on my local computer.

**Comment by DjebbZ** [comment]
Good to know, thanks for the pointer.

**Pull request merged**